pmlc/dialect/tile/builder.cc

```diff
@@ -769,6 +769,10 @@ TileBuilder::MakeProgram(StringRef name, const ProgramMutations &mutations,
   for (unsigned i = 0; i < returnOp.getNumOperands(); i++) {
     auto userValue = outputs[i];
     auto finalValue = returnOp.getOperand(i);
+    if (!finalValue.getDefiningOp()) {
+      IVLOG(2, "reached condition where return operand has no defining op");
+      userValue = MakePrimitiveOp("ident", {outputs[i]});
+    }
```

Comment:
Thanks for the explanation!

I've been looking into ways to add an `ident` op around any outputs that need copies of the input tensor once optimization passes are complete. One of the ways I've been exploring is to check (at the end of `MakeProgram`) whether each `ReturnOp` argument has a defining op, and wrap the argument with `ident` if that defining op is not found.

I've added an `IVLOG` to see where this condition is reached, and out of all the cc_tests, the only time it ever gets reached is during `ReshapeFold` (which is expected, because that's the only place where we are returning an input value that hasn't already got an `ident` wrapper around it). However, I know that my attempt to wrap the output with `ident` isn't working, since the `ident` statement isn't added when the condition is reached. This is a partial thing while I figure that out (or if you have any ideas/feedback, let me know)

Reply:
I think there were a bunch of keras backend tests that would trip this condition. @tzerrell might recall better which tests those might be (tests that would produce empty program bodies).

Reply:
I think there might have been a few other tests that happened to hit this, but the one that is attempting to test this is `TestBackendOps.testPassthrough`.

Reply:
Also possibly related: `testConstant`, `testUpdate`, `testSelfMult`, and `testDupOutputs`.